### PR TITLE
Adding Timeout Property to DurableHttpRequest

### LINKED ISSUE #1278: Add maximum and custom polling duration to DurableHttpRequest
This issue is pulled from a comment made in the ["Allow custom polling behavior in CallHttpAsync() for long-running endpoints"](https://github.com/Azure/azure-functions-durable-extension/issues/1183) issue.

The comments say:
"Second, it would be nice to be able to specify an overall maximum polling duration, after which it could throw a time out exception.

Third, from my experiments it seems that the original HTTP request will time out in just under 90 seconds. It would be nice to be able to control that timeout duration with another property on `DurableHttpRequest`."

Here are some solutions to this issue:
An additional property (e.g. `maximumPollingDuration`) can be added to `DurableHttpRequest` to keep track of an overall maximum polling duration.
 
An additional property (e.g. `timeOutDuration`) can be added to `DurableHttpRequest` to specify if there is a custom timeout duration for the original HTTP request.

### PR DESCRIPTION
**Pull request by bachuv**
These changes resolve #1278 by adding a `Timeout` property to `DurableHttpRequest`. This property defines the timeout for each individual HTTP request. This timeout applies to each asynchronous polling request too.

### CODE REVIEW

File: src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs
Comment (bachuv) [inline src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs:75]: To make this more accurate, it could be `catch (OperationCanceledException ex) when (cts.Token.IsCancellationRequested)` (added the commented out part). This way we know that the exception came when a cancel request was sent to `CancellationTokenSource`. The only issue with this is adding tests that mirror the action of cancelling a `CancellationTokenSource`. The `Mock<HttpMessageHandler>` Setup call takes `ItExpr.IsAny<CancellationToken>()` and the test would need the `CancellationTokenSource` to cancel the task, not the `CancellationToken`.

I saw that that `ItExpr.IsAny()` could evaluate conditions, but didn't see anything about sending in another object to manipulate. Let me know if there's a way to pass in a CancellationTokenSource using Moq.

This is how the `HttpMessageHandler` is mocked right now:
```
var handlerMock = new Mock<HttpMessageHandler>(MockBehavior.Strict);
handlerMock 
  .Protected()
  .Setup<Task<HttpResponseMessage>>("SendAsync", ItExpr.IsAny<HttpRequestMessage>(), ItExpr.IsAny<CancellationToken>())
  .ReturnsAsync(httpResponseMessage);

```

File: src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs
Comment (bachuv) [inline src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs:56]: Any suggestions on what to do if `DateTime.UtcNow` is greater than `durableHttpRequest.TimeoutExpirationDateTime`? Right now, it catches the exception that says that the `timeout` TimeSpan is invalid and resets the `timeout` value. Should it just throw a TimeoutException?

File: test/Common/DurableHttpTests.cs
Comment (bachuv) [inline test/Common/DurableHttpTests.cs:388]: Right now, the `CallHttpAsync` API throws a `FunctionFailedException` when it reaches a timeout, so that's why these are the Assert conditions that the test is checking. The test should be checking for a `TimeoutException` using `Assert.ThrowsAsync()`. Do you have any recommendations on how to propagate the `TimeoutException`? I went through the stack trace and it looks like it fails to deserialize the exception in TaskActivityDispatcher.cs, which I can't debug.

This is an example of the exception that's thrown when the function reaches the timeout:
```
Function 'Function1 (Orchestrator)' failed with an error. Reason: Microsoft.Azure.WebJobs.Extensions.DurableTask.FunctionFailedException: The activity function 'BuiltIn::HttpActivity' failed: "Failed to deserialize exception from TaskActivity: Unhandled exception while executing task: System.TimeoutException: A task was canceled.Reached user specified timeout: 00:00:10

```

File: src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs
Comment (ConnorMcMahon) [reply src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs:56]: So I can see an argument for one of two approaches:

1. When the customer sets the `timeout` property, it is absolute. If it expires before we hit the activity, don't even bother with the HTTP request. The good part of this approach is that it is straightforward to explain how it works. There are a couple of downsides. The big one is that the request could never even fire successfully due to internal Durable Task latency (such as a queue backlog). The other downside I see is that it doesn't give the customer much more usefulness other than just doing a Task.WhenAny() with a Timer task.

2. We only check for the timeout before we "schedule" the HTTP call. The downside is it is unclear what to set as the request cancellation timeout token to if the timeout has expired when we enter the HttpActivityShim.

Comment (ConnorMcMahon) [reply src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs:56]: @cgillum, what are your thoughts here?

File: src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs
Comment (ConnorMcMahon) [inline src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs:81]: Is there a reason we split this out into a helper method?

File: test/Common/DurableHttpTests.cs
Comment (ConnorMcMahon) [reply test/Common/DurableHttpTests.cs:388]: So the issue is for some reason TimeoutException is not being serialized properly... We could throw a different exception instead or handle serializing this time of exception more gracefully.

**Review by ConnorMcMahon — COMMENTED** [review]
Answered some questions and gave some feedback, but there is still some stuff that needs to be clarified.

File: src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs
Comment (bachuv) [reply src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs:81]: RunAsync() needs access to the DurableHttpRequest to figure out if there is a timeout or not.

File: src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableOrchestrationContext.cs
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableOrchestrationContext.cs:260]: There is no benefit to having a `catch` block that only does a `throw`. Let's remove the try/catch unless there is some way we can make it more useful.

File: src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableOrchestrationContext.cs
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableOrchestrationContext.cs:293]: Rather than relying on a cancellation token, would it make sense to instead use the minimum of either `fireAt` or `req.TimeoutExpirationDateTime` as the parameter to `CreateTimer`?

File: src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs
Comment (cgillum) [reply src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs:48]: +1, we don't want to set the timeout on the HTTP client since that could change the behavior for other, unrelated code.

File: src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs:42]: `CancellationTokenSource` implements `IDisposable`, so we need to wrap it in a `using` statement.

File: src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs
Comment (cgillum) [reply src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs:51]: +1, exceptions should only be used in exceptional cases - not for control flow.

File: src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs
Comment (cgillum) [reply src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs:56]: I'm not actually sure I understand the difference between those two options. However, it looks like we need to make another design decision about how the timeout should work. It seems there are two options, and maybe you were already alluding to this and I misunderstood.

1. Timeout is "absolute" in the sense that it represents a point in time. In this model, it would also include time spent in the work-item queue.
2. Timeout is relative in the sense that it does not have a specific expiration, but rather dictates how much time the HTTP client should wait for a response.

My gut tells me that (2) will be a more common scenario. In that case, we don't need to calculate any expiration time at all. We just pass the timeout value all the way to this point and use it in the call to `cts.CancelAfter()`. I worry that with (1) customers will have unpredictable and/or unexpected runtime behavior.

Thoughts?

File: test/Common/DurableHttpTests.cs
Comment (cgillum) [reply test/Common/DurableHttpTests.cs:388]: Let's try to figure out why it's not being serialized properly. If it's a problem for us then it's likely also a problem for our customers when they have activity functions that throw this exception.

**Review by cgillum — CHANGES_REQUESTED** [review]
Adding a bit of my own feedback here.

File: src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs
Comment (bachuv) [reply src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs:56]: I think it will be easier to document and maintain the second option (the timeout specifies the time the HTTP client should wait for a response). I agree that option 1 might be confusing for customers.

File: src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableOrchestrationContext.cs
Comment (ConnorMcMahon) [inline src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableOrchestrationContext.cs:699]: Unrelated to this PR, but can you file an in depth issue of your findings here of why we needed this workaround, so we can investigate whether this impacts other exceptions customers will throw and if we can fix it in a non-breaking way?

File: src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs
Comment (ConnorMcMahon) [inline src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs:44]: small nit:

We could make this a bit cleaner by limiting the introduction of the using block, and the try-catch block for `OperationCanceledException` into the case where `durableHttpRequest.Timeout` is non-null. It moderately improves readability by keeping the more general case less indented.

File: src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs
Comment (ConnorMcMahon) [inline src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs:82]: nit: This could probably be renamed `ConstructHttpRequestMessage()` or `BuildHttpRequestMessage()` or `ConvertToHttpRequestMessage()`, as we never really had a `HttpRequestMessage` in the first place,

**Review by ConnorMcMahon — APPROVED** [review]
Looks good. I have a small nit, but not a blocker. It's up to you whether you want to fix it before merging.

File: src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableOrchestrationContext.cs
Comment (cgillum) [reply src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableOrchestrationContext.cs:699]: This was my suggestion to Varshi for how to know when to throw `TimeoutException` instead of `FunctionFailedException`. We're not really working around anything AFAIK. Surfacing outer exceptions other than `FunctionFailedException` is not something we expose to our customers.

**Review by cgillum — APPROVED** [review]
Resetting my approval status. No blockers from me.

File: src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableOrchestrationContext.cs
Comment (ConnorMcMahon) [reply src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableOrchestrationContext.cs:689]: Makes sense.

I remember this being something we discussed around the V2 timeframe, but we missed our chance to tackle it before we locked in our breaking changes.

Probably worth keeping track of as a "breaking-change" issue that we can consider looking into if we make another breaking change release in the future (a ways off hopefully!).

**Pull request merged**
